### PR TITLE
Add grad check in test pipeline parallel fwd bwd

### PR DESCRIPTION
**Pull request by Aidyn-A**
Added a gradient comparison `test_pipeline_parallel_fwd_bwd.py`.
This PR is depends on #1384. 

cc @crcrpar @eqy

### CODE REVIEW

File: tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py [hunk 1]
```diff
@@ -31,33 +31,39 @@
...
+    loss.backward()
 
+    return float((data.sum()/global_batch_shape[0]).item()), model
 
 class PipelineParallelForwardBackwardTest(DistributedTestBase):
 
-    GLOBAL_BATCH_SIZE = 16
-    MICRO_BATCH_SIZE = 2
+    GLOBAL_BATCH_SIZE = 32
+    MICRO_BATCH_SIZE = 1
```
Comment (crcrpar) [inline tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py:66]: [question] Could you tell me why these changes are needed?
(I admit that these values are magic number though)

Comment (Aidyn-A) [reply tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py]: Sorry, I played with these numbers to make sure, that everything works fine and I forgot to change them to the original values. I reverted these changes in [d8e7171](https://github.com/NVIDIA/apex/pull/1386/commits/d8e71716026bf61daf5484653498228e7b18ddb4).

File: tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py [hunk 2]
```diff
@@ -170,11 +179,24 @@ def _forward_backward_test_impl(
                 total_layers = pipeline_model_parallel_world_size
                 if virtual_pipeline_model_parallel_size is not None:
                     total_layers *= virtual_pipeline_model_parallel_size
-                target_loss = get_target_loss(global_batch_shape, hidden_size, total_layers)
+                target_loss, target_model = get_target_loss_and_model(global_batch_shape, hidden_size, total_layers)
 
                 for loss_item in loss:
                     x = loss_item['avg']
-                    torch.testing.assert_close(x / microbatch_size, target_loss*torch.ones_like(x))
+                    torch.testing.assert_close(x[0].cpu() / microbatch_size, target_loss)
```
Comment (crcrpar) [inline tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py:186]: QQ: what's the `x`'s shape?

**Review by crcrpar — APPROVED** [review]

File: tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py (hunk 2 above)
Comment (eqy) [reply tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py:186]: Should be a scalar "tensor"

File: tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py [hunk 3]
```diff
@@ -170,11 +179,24 @@ def _forward_backward_test_impl(
...
+
+                if not forward_only:
+                    for vm_id, model_module in enumerate(model):
+                        params = list(model_module.parameters())
+                        rank = params[0].get_device()
+                        offset = pipeline_model_parallel_world_size
+                        param_id = (rank * offset) // self.world_size + vm_id * offset
```
Comment (eqy) [inline tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py:193]: Do we want `self.world_size` or would just `pipeline_model_parallel_world_size` be ok?

File: tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py (hunk 2 above)
Comment (Aidyn-A) [reply tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py:186]: I do this because there is a mismatch in sizes:

`x.size()` is `torch.Size([1])`
`target_loss.size()` is `torch.Size([])`

File: tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py (hunk 3 above)
Comment (Aidyn-A) [reply tests/L0/run_transformer/test_pipeline_parallel_fwd_bwd.py:193]: Agree, a better expression would be `param_id = rank // data_parallel_size + vm_id * offset`. I will commit the change when other conversations are resolved.

**Review by crcrpar — APPROVED** [review]
